### ISSUE TITLE
[question] Nesting sound effects.

### ISSUE THREAD

**Issue by metroidchild** [kcat/openal-soft#57]
In for example DAW's you can easily add multiple effects in a chain and then be on your way, what I'd like to try and make is a reverb with an echo like effect as the source (just to figure out how this can be achieved).

The basic chain (in very crude pseudo code) would look something like this:
`reverb( loop( sound ) );`
But due to how effects work in openal-soft I suspect it'll look something like:

```
effect0 = LoadEffect(&reverb);
effect1 = LoadEffect(&echo);

slot = 0;
alGenAuxiliaryEffectSlots(1, &slot);

alAuxiliaryEffectSloti(slot, AL_EFFECTSLOT_EFFECT, effect1);
alAuxiliaryEffectSloti(slot, AL_EFFECTSLOT_EFFECT, effect0);
```

...or some other thing, can someone give me some guidance or is this impossible to do?

**Comment by kcat** [comment]
The ```AL_EFFECT_EAXREVERB``` effect contains echo parameters (```AL_EAXREVERB_ECHO_TIME``` and ```AL_EAXREVERB_ECHO_DELAY```). It may be worth playing with them to see if it gets you what you're looking for.

Alternatively, you can simply play the same sound multiple times with a delay between the play calls (you'll need multiple sources to cycle through, but they can all use the same buffer). This may not be quite so good though, depending on the desired echo time (OpenAL Soft's update time can be relatively granular).

Otherwise, the EFX API doesn't currently allow chaining individual effects. The current Git version is suited to allowing it, if an API extension is added to do it.

**Comment by metroidchild** [comment]
Thanks for the answer! I might just try and see if I can botch up something that will do that, after reading some of the source code it seems hard but doable

**Comment by Member1221** [comment]
Has there been any progress towards getting this working?

**Comment by kcat** [comment]
Commit 194e7ff815d744830c8aaedfb0d32ed3c3d01645 adds an in-progress extension to do this. Be aware that in-progress extensions are subject to change and should not be considered "production ready", but it should be enough to test and provide feedback on. As it is currently, effects can be chained like this:
```
alAuxiliaryEffectSloti(effectslot0, AL_EFFECTSLOT_TARGET_SOFT, effectslot1);
```
In this case, the output of effectslot0 is added to the input of effectslot1. So if `effectslot0` has an echo effect loaded, and `effectslot1` has a reverb effect, the echoes will be used as input to the reverb (a source would not need to send to effectslot1 explicitly, just to effectslot0 is enough).

Some things to note: each effect slot can only have one target (though an effect slot can act as the target for multiple other effect slots and sources). You can't delete an effect slot that is currently the target of another. It's an error to create a circular chain.

**Comment by Member1221** [comment]
My use of it can be found here https://github.com/PolyplexEngine/libpp/blob/master/source/polyplex/core/audio/effect.d#L124
Thanks for working on this!

**Comment by kunal-vzw** [comment]
Any update on when this will be available in the tagged/release versions of OpenAL Soft? Thanks

**Comment by kcat** [comment]
Presuming there's no significant issues with [the proposal](http://openal.org/pipermail/openal/2020-April/000745.html), it will probably be in the next release. Which itself could be in a month or two.

**Comment by kcat** [comment]
Closing since it's been available since 1.21.
